tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+            selfrows = np.around(np.linspace(self.bbox[1], self.bbox[3],
+                       self.shape[0], endpoint=False)[::-1],precision)
+            selfcols = np.around(np.linspace(self.bbox[0], self.bbox[2],
+                       self.shape[1], endpoint=False), precision)
+            new_bbox = (selfcols[nz_ix[2]], selfrows[nz_ix[1]],
+                        selfcols[nz_ix[3]], selfrows[nz_ix[0]])
+            # set self.bbox to clipped bbox
+            self.set_bbox(new_bbox, **kwargs)
+        else:
+            # if inplace is False, return the clipped data
+            return data[nz_ix[0]:nz_ix[1], nz_ix[2]:nz_ix[3]]
+
+    def set_bbox(self, new_bbox, precision=7): 
+        """
+        Set the bounding box of the class instance (self.bbox) and clip
+        all grids to the new bbox.
+
+        Parameters
+        ----------
+        new_bbox : tuple
+                   New bbox to use (xmin, ymin, xmax, ymax)
+        precision : int
+                    Precision to use when matching geographic coordinates.
+        fillna : bool
+                 Whether to fill nulls.
+        fill : int or float
+               Fill value to use.
+        """
+        # round new bbox to proper precision
+        new_bbox = np.around(new_bbox, precision)
+
+        # construct arrays representing coordinates of existing grids
+        selfrows = np.around(np.linspace(self.bbox[1], self.bbox[3],
+                   self.shape[0], endpoint=False)[::-1], precision)
+        selfcols = np.around(np.linspace(self.bbox[0], self.bbox[2],
+                   self.shape[1], endpoint=False), precision)
+
+        # construct arrays representing coordinates of new grid
+        rows = (pd.Series(selfrows, index=selfrows)
+                .loc[new_bbox[3]:new_bbox[1]].values)
+        cols = (pd.Series(selfcols, index=selfcols)
+                .loc[new_bbox[0]:new_bbox[2]].values)
+
+        # clip existing grids to new bbox
+        for i in self.gridlist:
+            data = pd.DataFrame(getattr(self, i),
+                                index=selfrows, columns=selfcols)
+            data = (data.reindex(rows).reindex_axis(cols, axis=1)
+                        .fillna(self.nodata[i]).values)
+            setattr(self, i, data)
+        
+        # set class attributes
+        self.bbox = tuple(new_bbox)
+        self.shape = tuple([len(rows), len(cols)])
+
+    def set_nodata(self, data_name, new_nodata, old_nodata=None):
+        """
+        Change nodata value of dataset.
+
+        Parameters
+        ----------
+        data_name : string
+                    Attribute name of dataset to change
+        new_nodata : int or float
+                     New nodata value to use
+        old_nodata : int or float (optional)
+                     If none provided, defaults to self.nodata[data_name]
+        """
+
+        if old_nodata is None:
+            old_nodata = self.nodata[data_name]
+        data = getattr(self, data_name)
```

Comment:
why not just `data = self.data_name`?  This syntax is used elsewhere as well and I'm not sure why.